File: public/apps/chmod_calculator/chmod_calculator.js
document.addEventListener('DOMContentLoaded', () => {
  const checkboxes = document.querySelectorAll('input[type="checkbox"]');
  const octalInput = document.getElementById('octal');
  const symbolicInput = document.getElementById('symbolic');
  const themeSwitch = document.getElementById('theme-switch');
  const body = document.body;

  const updatePermissions = () => {
    const data = {
      owner_read: document.getElementById('owner-read').checked,
      owner_write: document.getElementById('owner-write').checked,
      owner_execute: document.getElementById('owner-execute').checked,
      group_read: document.getElementById('group-read').checked,
      group_write: document.getElementById('group-write').checked,
      group_execute: document.getElementById('group-execute').checked,
      public_read: document.getElementById('public-read').checked,
      public_write: document.getElementById('public-write').checked,
      public_execute: document.getElementById('public-execute').checked,
    };

    fetch('http://localhost:3000/calculate', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(data),
    })
    .then(response => {
      if (!response.ok) {
        return response.json().then(err => { throw new Error(err.error); });
      }
      return response.json();
    })
    .then(data => {
      octalInput.value = data.octal;
      symbolicInput.value = data.symbolic;
    })
    .catch(error => {
      console.error('Error:', error);
      octalInput.value = 'Error';
      symbolicInput.value = 'Error';
    });
  };

  const resetPermissions = () => {
    checkboxes.forEach(checkbox => checkbox.checked = false);
    octalInput.value = '';
    symbolicInput.value = '';
    updatePermissions();
  };

  themeSwitch.addEventListener('change', () => {
    if (themeSwitch.checked) {
      body.classList.remove('light-theme');
      body.classList.add('dark-theme');
    } else {
      body.classList.remove('dark-theme');
      body.classList.add('light-theme');
    }
  });

  checkboxes.forEach(checkbox => {
    checkbox.addEventListener('change', updatePermissions);
  });

  document.querySelector('button').addEventListener('click', resetPermissions);

  // Initial calculation
  updatePermissions();
});

File: public/apps/chmod_calculator/chmod_calculator.py
import sys
import json

def calculate_permissions(data):
    owner = (4 if data['owner_read'] else 0) + (2 if data['owner_write'] else 0) + (1 if data['owner_execute'] else 0)
    group = (4 if data['group_read'] else 0) + (2 if data['group_write'] else 0) + (1 if data['group_execute'] else 0)
    public = (4 if data['public_read'] else 0) + (2 if data['public_write'] else 0) + (1 if data['public_execute'] else 0)

    octal = f"{owner}{group}{public}"
    symbolic = (f"{'r' if data['owner_read'] else '-'}{'w' if data['owner_write'] else '-'}{'x' if data['owner_execute'] else '-'}"
                f"{'r' if data['group_read'] else '-'}{'w' if data['group_write'] else '-'}{'x' if data['group_execute'] else '-'}"
                f"{'r' if data['public_read'] else '-'}{'w' if data['public_write'] else '-'}{'x' if data['public_execute'] else '-'}")
    
    return octal, symbolic

if __name__ == "__main__":
    try:
        input_data = json.loads(sys.stdin.read())
        octal, symbolic = calculate_permissions(input_data)
        result = {"octal": octal, "symbolic": symbolic}
        print(json.dumps(result))  # Print result to stdout
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)
